addons/purchase/models/account_invoice.py

```diff
@@ -52,10 +52,12 @@ def _onchange_purchase_auto_complete(self):
 
         # Copy data from PO
         invoice_vals = self.purchase_id.with_company(self.purchase_id.company_id)._prepare_invoice()
-        invoice_vals['currency_id'] = self.invoice_line_ids and self.currency_id or invoice_vals.get('currency_id')
+        new_currency_id = self.invoice_line_ids and self.currency_id or invoice_vals.get('currency_id')
         del invoice_vals['ref']
         del invoice_vals['company_id']  # avoid recomputing the currency
+        del invoice_vals['currency_id']
         self.update(invoice_vals)
+        self.update({'currency_id': new_currency_id})  # update currency at the end to avoid it being erased because of the computations triggered by the change of previous attributes
```

Comment:
```suggestion
        del invoice_vals['ref']
        del invoice_vals['company_id']  # avoid recomputing the currency
        self.update(invoice_vals)
        self.currency_id = self.invoice_line_ids and self.currency_id or invoice_vals.get('currency_id')
```

Reply:
Hello,
Here are my remarks :
Is it not easier to simply set the currency_id after the update() ?
Also, this should maybe be done in an l10n_in override ? @jco-odoo  ?

Reply:
@jbw-odoo  It depends.  I do not see something in l10n_in that would let currency_id and journal_id work differently than in generic as @bebo-odoo seems to pretend in his commit message.

Reply:
It's due to this function in the override of purchase_order in l10n_in_purchase that add the journal_id to the list of fields to update. Any change in the journal_id recomputes the currency_id, erasing the previously set value
![image](https://github.com/odoo/odoo/assets/130741341/2e4c5e9e-d340-4115-bf43-c21cc579ffea)

Reply:
@bebo-odoo But that should not be used anymore...

Reply:
@jco-odoo I'm sorry, what shouldn't be used? :sweat_smile:

Reply:
> @jco-odoo I'm sorry, what shouldn't be used? 😅

I mean that the functionality of the l10n_in_journal_id is deprecated normally, but it is true it is still in the code.

Reply:
@jco-odoo The thing is, they both (journal_id and currency_id) get computed at the same time as they are both in the vals
![image](https://github.com/odoo/odoo/assets/130741341/535e3081-a781-431b-8aa4-a25129750ad8)
So it's just a matter of chance that the journal_id gets updated after, it could have been the opposite if Python sorted their dict differently when looping over it.
From the client perspective, it doesn't make sense that upon autocompletion the lines don't get the same currency as the autocompleted PO as it is the case with other localizations, hence why I think that the newly set currency_id should be prioritized over the one computed by the journal_id

Reply:
@jbw-odoo updated according to your suggestion, you're right it's just simpler this way, was too focused on updating the currency using the update function...
In my opinion it's better to write it here rather as in the l10n_in override, as it fixes the problem for any localization that changes a field that would override the value of the currency.